File: packages/client/src/siwx.ts
/**
 * Sign-In-With-X (SIWx) Client SDK
 *
 * Implements CAIP-122 standard wallet-based identity assertions.
 * This module allows clients to prove control of a wallet for authentication.
 */

// ============================================================================
// Type Definitions
// ============================================================================

/**
 * Network identifier in CAIP-2 format
 */
export type Network = `${string}:${string}`;

/**
 * Supported signature schemes for SIWx
 */
export type SignatureScheme =
  | "eip191"
  | "eip712"
  | "eip1271"
  | "eip6492"
  | "siws"
  | "sep10";

/**
 * SIWx payload containing the message fields and signature
 */
export interface SIWxPayload {
  /** Domain that requested the signature */
  domain: string;
  /** Address of the signer */
  address: string;
  /** Optional statement */
  statement?: string;
  /** URI of the resource */
  uri: string;
  /** SIWx version */
  version: string;
  /** Chain ID in CAIP-2 format */
  chainId: string;
  /** Optional nonce */
  nonce?: string;
  /** Optional time when the message was issued */
  issuedAt?: string;
  /** Time when the message expires (required) */
  expirationTime: string;
  /** Time before which the message is not valid */
  notBefore?: string;
  /** Optional request identifier */
  requestId?: string;
  /** List of resources */
  resources?: string[];
  /** Signature over the message */
  signature: string;
}

/**
 * Signer interface for signing SIWx messages
 */
export interface SIWxSigner {
  /** Sign a SIWx payload and return the base64-encoded envelope */
  signPayload(payload: SIWxPayload): Promise<string>;
}

// ============================================================================
// Client-Side Functions
// ============================================================================

/**
 * Get the chain name from a CAIP-2 chain ID
 */
function getChainName(chainId: string): string {
  if (chainId.startsWith("eip155:")) {
    return "Ethereum";
  }
  if (chainId.startsWith("solana:")) {
    return "Solana";
  }
  // Extract namespace for unknown chains
  const namespace = chainId.split(":")[0];
  return namespace.charAt(0).toUpperCase() + namespace.slice(1);
}

/**
 * Create a CAIP-122 formatted SIWx message string from a payload
 *
 * @param payload - SIWx payload (signature field is ignored)
 * @returns CAIP-122 formatted message string
 *
 * @example
 * ```typescript
 * const message = createSIWxMessage({
 *   domain: "example.com",
 *   address: "0x1234...",
 *   uri: "https://example.com/resource",
 *   version: "1",
 *   chainId: "eip155:1",
 *   expirationTime: new Date(Date.now() + 3600000).toISOString(), // 1 hour from now
 *   signature: "", // ignored
 * });
 * ```
 */
export function createSIWxMessage(
  payload: Omit<SIWxPayload, "signature"> | SIWxPayload
): string {
  const chainName = getChainName(payload.chainId);

  const lines: string[] = [
    `${payload.domain} wants you to sign in with your ${chainName} account:`,
    payload.address,
    "",
  ];

  if (payload.statement) {
    lines.push(payload.statement);
    lines.push("");
  }

  lines.push(`URI: ${payload.uri}`);
  lines.push(`Version: ${payload.version}`);
  lines.push(`Chain ID: ${payload.chainId}`);

  if (payload.nonce) {
    lines.push(`Nonce: ${payload.nonce}`);
  }

  if (payload.issuedAt) {
    lines.push(`Issued At: ${payload.issuedAt}`);
  }

  lines.push(`Expiration Time: ${payload.expirationTime}`);

  if (payload.notBefore) {
    lines.push(`Not Before: ${payload.notBefore}`);
  }

  if (payload.requestId) {
    lines.push(`Request ID: ${payload.requestId}`);
  }

  if (payload.resources && payload.resources.length > 0) {
    lines.push("Resources:");
    for (const resource of payload.resources) {
      lines.push(`- ${resource}`);
    }
  }

  return lines.join("\n");
}

/**
 * Encode a signed SIWx envelope as a base64 header value
 *
 * @param message - The CAIP-122 formatted message string
 * @param signature - The signature over the message
 * @returns Base64-encoded header value
 *
 * @example
 * ```typescript
 * const header = encodeSIWxHeader(message, signature);
 * fetch(url, {
 *   headers: { "SIGN-IN-WITH-X": header }
 * });
 * ```
 */
export function encodeSIWxHeader(message: string, signature: string): string {
  const envelope = { message, signature };
  return btoa(JSON.stringify(envelope));
}

/**
 * Prepare a SIWx payload for signing
 *
 * This is a convenience function that creates the message and provides
 * a helper to encode the final header after signing.
 *
 * @param payload - SIWx payload (without signature)
 * @returns Object containing the message to sign and a function to create the header
 *
 * @example
 * ```typescript
 * const { message, createHeader } = prepareSIWxForSigning(payload);
 * const signature = await signer.signMessage(message);
 * const header = createHeader(signature);
 * ```
 */
export function prepareSIWxForSigning(
  payload: Omit<SIWxPayload, "signature">
): {
  message: string;
  createHeader: (signature: string) => string;
} {
  const message = createSIWxMessage(payload);

  return {
    message,
    createHeader: (signature: string) => encodeSIWxHeader(message, signature),
  };
}

File: python/aimo_network/client/siwx.py
"""Sign-In-With-X (SIWx) Client SDK.

Implements CAIP-122 standard wallet-based identity assertions.
This module allows clients to prove control of a wallet for authentication.
"""

from __future__ import annotations

import base64
import json
from dataclasses import dataclass, field
from typing import Callable


@dataclass
class SIWxPayload:
    """SIWx payload containing the message fields and signature."""

    domain: str
    address: str
    uri: str
    version: str
    chain_id: str
    expiration_time: str
    statement: str | None = None
    nonce: str | None = None
    issued_at: str | None = None
    not_before: str | None = None
    request_id: str | None = None
    resources: list[str] = field(default_factory=list)
    signature: str = ""


def _get_chain_name(chain_id: str) -> str:
    """Get the chain name from a CAIP-2 chain ID."""
    if chain_id.startswith("eip155:"):
        return "Ethereum"
    if chain_id.startswith("solana:"):
        return "Solana"
    namespace = chain_id.split(":")[0]
    return namespace[0].upper() + namespace[1:]


def create_siwx_message(payload: SIWxPayload) -> str:
    """Create a CAIP-122 formatted SIWx message string from a payload.

    Args:
        payload: SIWx payload (signature field is ignored).

    Returns:
        CAIP-122 formatted message string.
    """
    chain_name = _get_chain_name(payload.chain_id)

    lines: list[str] = [
        f"{payload.domain} wants you to sign in with your {chain_name} account:",
        payload.address,
        "",
    ]

    if payload.statement:
        lines.append(payload.statement)
        lines.append("")

    lines.append(f"URI: {payload.uri}")
    lines.append(f"Version: {payload.version}")
    lines.append(f"Chain ID: {payload.chain_id}")

    if payload.nonce:
        lines.append(f"Nonce: {payload.nonce}")

    if payload.issued_at:
        lines.append(f"Issued At: {payload.issued_at}")

    lines.append(f"Expiration Time: {payload.expiration_time}")

    if payload.not_before:
        lines.append(f"Not Before: {payload.not_before}")

    if payload.request_id:
        lines.append(f"Request ID: {payload.request_id}")

    if payload.resources:
        lines.append("Resources:")
        for resource in payload.resources:
            lines.append(f"- {resource}")

    return "\n".join(lines)


def encode_siwx_header(message: str, signature: str) -> str:
    """Encode a signed SIWx envelope as a base64 header value.

    Args:
        message: The CAIP-122 formatted message string.
        signature: The signature over the message.

    Returns:
        Base64-encoded header value.
    """
    envelope = {"message": message, "signature": signature}
    return base64.b64encode(json.dumps(envelope).encode()).decode()


@dataclass
class PreparedSIWx:
    """Result of preparing a SIWx payload for signing."""

    message: str
    create_header: Callable[[str], str]


def prepare_siwx_for_signing(payload: SIWxPayload) -> PreparedSIWx:
    """Prepare a SIWx payload for signing.

    This is a convenience function that creates the message and provides
    a helper to encode the final header after signing.

    Args:
        payload: SIWx payload (signature field is ignored).

    Returns:
        Object containing the message to sign and a function to create the header.
    """
    message = create_siwx_message(payload)
    return PreparedSIWx(
        message=message,
        create_header=lambda signature: encode_siwx_header(message, signature),
    )
